### PR TITLE
add default resource requirements to init-containers and containers of a pod

### PR DESCRIPTION
**Pull request by jkandasa**
# Changes
allow users to update default resource requirements to a init-containers and containers of a pod via `config-defaults` (or `CONFIG_DEFAULTS_NAME`) config map under `default-container-resource-requirements`.

The container names are dynamic, can be any name. some of the well know container names are: `place-scripts`, `prepare`, `working-dir-initializer`.

Also supports the container name which starts with prefix, examples: `scripts-xxx`, `sidecar-scripts-xxx`, etc.,
You have to include `prefix-` key in-front of the container name.
Example:
* `scripts-xxx` => actual key: `prefix-scripts`
* `sidecar-scripts-xxx` => actual key: `prefix-sidecar-scripts`

In addition to that there is a static key called `default`, which covers all the left out containers. If the `default` resource requirements specified and a container(or init-container) without resource requirements will be updated with the `default` values.

> ### NOTE: All the resource requirements are applied to `init-containers` and `containers` only if the existing resource requirements are empty.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-defaults
  namespace: tekton-pipelines
data:
  default-container-resource-requirements: |
    place-scripts: # updates resource requirements of a 'place-scripts' container
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m"
  
    prepare: # updates resource requirements of a 'prepare' container
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "256Mi"
        cpu: "500m"
  
    working-dir-initializer: # updates resource requirements of a 'working-dir-initializer' container
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "512Mi"
        cpu: "500m"
  
    prefix-scripts: # updates resource requirements of containers which starts with 'scripts-'
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m"
  
    prefix-sidecar-scripts: # updates resource requirements of containers which starts with 'sidecar-scripts-'
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m"
  
    default: # updates resource requirements of init-containers and containers which has empty resource resource requirements
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "256Mi"
        cpu: "500m"
```

* fixes: https://github.com/tektoncd/pipeline/issues/2933
* fixes: https://github.com/tektoncd/pipeline/issues/7058
* fixes: https://github.com/tektoncd/pipeline/issues/6974
* [SRVKP-2961](https://issues.redhat.com/browse/SRVKP-2961) - resource requests and limits not applied to init and sidecar container in pipelines
* [SRVKP-3140]( https://issues.redhat.com/browse/SRVKP-3140) - Resource limits and requests defined in Task stepTemplate are not applied to the related Pod's initContainers

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
[Bug fix]: takes default values of a resource requirements from a config map and updates to a init-container and container resource requirements value, if the value is not present
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
This looks more like a feature than a bug. Isn't it so @jkandasa ?

**Comment by jimmyjones2** [comment]
Can some sane defaults be applied as standard to init-containers (that can be tweaked with the ConfigMap if necessary)? Or maybe only if a ResourceQuota is present in the namespace? See #7058 

I can imagine lots of people with ResourceQuotas are also not cluster admins, and cluster admins don't like tweaking things, especially if they could affect multiple users...

**Review by vdemeester — APPROVED** [review]
cc @tektoncd/core-maintainers

**Comment by Yongxuanzhang** [comment]
Is it ok to add example and docstrings here: https://github.com/tektoncd/pipeline/blob/main/config/config-defaults.yaml
Basically add a short version of the example in your PR description

File: pkg/pod/pod.go
Comment (Yongxuanzhang) [inline pkg/pod/pod.go:531]: This is confusing, we're not updating container name here. same for other docstrings

**Comment by Yongxuanzhang** [comment]
Thanks for the fix! The code lgtm, I just wonder if we could have improve the documentation/docstrings 🙏

**Comment by chitrangpatel** [comment]
@jkandasa we're hoping to get this in for this release. PTAL and let us know if we can help.

**Comment by jkandasa** [comment]
> @jkandasa we're hoping to get this in for this release. PTAL and let us know if we can help.

Thanks, I will work on this as soon as possible.

**Comment by vdemeester** [comment]
@Yongxuanzhang @chitrangpatel Updated some docstrings, and added an example in the configmap, and documentation.

**Comment by Yongxuanzhang** [comment]
./config/config-defaults.yaml fail the lint test 
https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_pipeline/7003/pull-tekton-pipeline-build-tests/1747247928986046464/build-log.txt
```
./config/config-defaults.yaml
  102:6     error    trailing spaces  (trailing-spaces)
  110:6     error    trailing spaces  (trailing-spaces)
  118:6     error    trailing spaces  (trailing-spaces)
  126:6     error    trailing spaces  (trailing-spaces)
  134:6     error    trailing spaces  (trailing-spaces)
```

**Pull request merged**
